main.py

```diff
@@ -260,49 +360,9 @@ def main(cl_arguments):
                           to_train, opt_params, schd_params,
                           args.shared_optimizer, args.load_model, phase="main")
 
-    # Select model checkpoint from main training run to load
-    if not args.do_target_task_training:
-        log.info("In strict mode because do_target_task_training is off. "
-                 "Will crash if any tasks are missing from the checkpoint.")
-        strict = True
-    else:
-        strict = False
-
-    if args.do_target_task_training and not args.allow_reuse_of_pretraining_parameters:
-        # If we're training models for evaluation, which is always done from scratch with a fresh
-        # optimizer, we shouldn't load parameters for those models.
-        # Usually, there won't be trained parameters to skip, but this can happen if a run is killed
-        # during the do_target_task_training phase.
-        task_names_to_avoid_loading = [task.name for task in target_tasks]
-    else:
-        task_names_to_avoid_loading = []
-
-    if not args.load_eval_checkpoint == "none":
-        # This is to load a particular eval checkpoint.
-        log.info("Loading existing model from %s...", args.load_eval_checkpoint)
-        load_model_state(model, args.load_eval_checkpoint,
-                         args.cuda, task_names_to_avoid_loading, strict=strict)
-    else:
-        # Look for eval checkpoints (available only if we're restoring from a run that already
-        # finished), then look for training checkpoints.
-
-        if args.transfer_paradigm == "finetune":
-            # Save model so we have a checkpoint to go back to after each task-specific finetune.
-            model_state = model.state_dict()
-            model_path = os.path.join(args.run_dir, "model_state_untrained_prefinetune.th")
-            torch.save(model_state, model_path)
-
-        best_path = get_best_checkpoint_path(args.run_dir)
-        if best_path:
-            load_model_state(model, best_path, args.cuda, task_names_to_avoid_loading,
-                             strict=strict)
-        else:
-            assert_for_log(args.allow_untrained_encoder_parameters,
-                           "No best checkpoint found to evaluate.")
-            log.warning("Evaluating untrained encoder parameters!")
-
     # Train just the task-specific components for eval tasks.
     if args.do_target_task_training:
+        task_names_to_avoid_loading, strict = setup_target_task_training(args, target_tasks, model)
```

Comment:
Actually I think it would be correct if we just moved this outside of the loop?

Reply:
It's not in a loop?